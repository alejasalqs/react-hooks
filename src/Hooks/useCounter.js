import { useState } from "react";
// Un custom hook es igual que una funcion normal de react
export const useCounter = (initialState = 10) => {
  // vamos a centralizar la logica de un contador en este hook
  const [state, setState] = useState(initialState);

  const increment = (factor = 1) => {
    setState(state + factor);
  };

  const decrement = (factor = 1) => {
    setState(state - factor);
  };

  const resetCounter = () => {
    setState(initialState);
  };
  // en este caso se retorna un objeto {} o arreglo []
  return {
    state,
    increment,
    decrement,
    resetCounter,
  };
};
