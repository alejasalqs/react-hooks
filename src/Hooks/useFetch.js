import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  const isMounted = useRef(true);

  useEffect(() => {
    // Solo se ejecuta la primera vez que se llama el componente
    return () => {
      // Cuando el componente se desmonta cancela se ejecuta esto
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });
    // aqui no se puede usar async/await
    fetch(url)
      .then((resp) => {
        resp.json().then((data) => {
          if (isMounted.current) {
            // Si el componente esta montado ejecuta
            setState({
              loading: false,
              error: null,
              data,
            });
          } else {
            console.log("setState no se llama");
          }
        });
      })
      .catch((e) => {
        setState({
          data: null,
          loading: false,
          error: "No se pudo cargar la informaci√≥n",
        });
      });
  }, [url]); // Se ejecuta cuando el URL cambia

  return state;
};
