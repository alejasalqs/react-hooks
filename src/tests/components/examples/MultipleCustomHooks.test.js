import React from "react";
import Adapter from "enzyme-adapter-react-16";
const { shallow, configure } = require("enzyme");
import { MultipleCustomHooks } from "../../../components/examples/MultipleCustomHooks";
import { useFetch } from "../../../Hooks/useFetch";
import { useCounter } from "../../../Hooks/useCounter";
// Esto lo que hace es una implementacion falsa del archivo que estamos llamando
// Con solo declrarlo ejecuta la accion
jest.mock("../../../Hooks/useFetch");
jest.mock("../../../Hooks/useCounter");

configure({ adapter: new Adapter() });

describe("Pruebas en el archivo <MultipleCustomHooks />", () => {
  // Como esto no cambia para todos los test case lo definimos en el describe
  /*useCounter.mockReturnValue({
    state: 10,
    increment: () => {},
  });*/ // Daba error entonces lo instancie en cada test

  test("Debe de mostrarse correctamente ", () => {
    // Instanciamos el useFetch para evitar errores
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    useCounter.mockReturnValue({
      state: 10,
      increment: () => {},
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar la infomaciÃ³n ", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "Alejandro",
          quote: "Hola Mundo",
        },
      ],
      loading: false,
      error: null,
    });

    useCounter.mockReturnValue({
      state: 10,
      increment: () => {},
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    // Busca un elemento por la clase
    expect(wrapper.find(".alert").exists()).toBe(false);
    expect(wrapper.find(".mb-0").text().trim()).toBe("Hola Mundo");

    //Busca un elemento por html tag
    expect(wrapper.find("footer").text().trim()).toBe("Alejandro");
  });
});
