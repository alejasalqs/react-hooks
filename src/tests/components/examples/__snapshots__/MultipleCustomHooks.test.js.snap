// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el archivo <MultipleCustomHooks /> Debe de mostrar la infomaci√≥n  1`] = `ShallowWrapper {}`;

exports[`Pruebas en el archivo <MultipleCustomHooks /> Debe de mostrarse correctamente  1`] = `ShallowWrapper {}`;
